tests/components/template/test_light.py

```diff
@@ -154,10 +158,165 @@
+            "{{ 255 }}",
+        ),
+        (
+            "supports_transition_template",
+            "supports_transition",
+            "{{ True }}",
+            "{{ True }}",
+        ),
+        (
+            "temperature_template",
+            "temperature",
+            "{{ 255 }}",
+            "{{ 255 }}",
+        ),
+        (
+            "white_value_template",
+            "white_value",
+            "{{ 255 }}",
+            "{{ 255 }}",
+        ),
+        (
+            "hs_template",
+            "hs",
+            "{{ (255, 255) }}",
+            "{{ (255, 255) }}",
+        ),
+        (
+            "color_template",
+            "hs",
+            "{{ (255, 255) }}",
+            "{{ (255, 255) }}",
+        ),
+    ],
+)
+async def test_legacy_to_modern_config(
+    hass: HomeAssistant, old_attr: str, new_attr: str, template: str, expected: str
+) -> None:
+    """Test the conversion of legacy template to modern template."""
+    config = {
+        "foo": {
+            "friendly_name": "foo bar",
+            "unique_id": "foo-bar-light",
+            "icon_template": "{{ 'mdi.abc' }}",
+            "entity_picture_template": "{{ 'mypicture.jpg' }}",
+            "availability_template": "{{ 1 == 1 }}",
+            old_attr: template,
+            **OPTIMISTIC_ON_OFF_LIGHT_CONFIG,
+        }
+    }
+    altered_configs = rewrite_legacy_to_modern_conf(hass, config)
+
+    assert len(altered_configs) == 1
+
+    altered_config = altered_configs[0]
+
+    assert "object_id" in altered_config
+    assert altered_config["object_id"] == "foo"
+
+    assert "friendly_name" not in altered_config
+    assert "name" in altered_config
+    assert altered_config["name"].template == "foo bar"
+
+    assert "icon_template" not in altered_config
+    assert "icon" in altered_config
+    assert altered_config["icon"].template == "{{ 'mdi.abc' }}"
+
+    assert "entity_picture_template" not in altered_config
+    assert "picture" in altered_config
+    assert altered_config["picture"].template == "{{ 'mypicture.jpg' }}"
+
+    assert "availability_template" not in altered_config
+    assert "availability" in altered_config
+    assert altered_config["availability"].template == "{{ 1 == 1 }}"
+
+    assert "unique_id" in altered_config
+    assert altered_config["unique_id"] == "foo-bar-light"
+
+    assert "turn_on" in altered_config
+    assert altered_config["turn_on"] == {
+        "service": "test.automation",
+        "data_template": {
+            "action": "turn_on",
+            "caller": "{{ this.entity_id }}",
+        },
+    }
+
+    assert "turn_off" in altered_config
+    assert altered_config["turn_off"] == {
+        "service": "test.automation",
+        "data_template": {
+            "action": "turn_off",
+            "caller": "{{ this.entity_id }}",
+        },
+    }
+
+    assert old_attr not in altered_config
+    assert new_attr in altered_config
+    assert altered_config[new_attr].template == expected
```

Comment:
I suggest to compare the entire dict instead of selected keys, it's IMO much easier to read and also guards against future changes

```suggestion
    assert [
        {
            "availability": Template("{{ 1 == 1 }}", hass),
            "icon": Template("{{ 'mdi.abc' }}", hass),
            "name": Template("foo bar", hass),
            "object_id": "foo",
            "picture": Template("{{ 'mypicture.jpg' }}", hass),
            "turn_off": {
                "data_template": {
                    "action": "turn_off",
                    "caller": "{{ this.entity_id }}",
                },
                "service": "test.automation",
            },
            "turn_on": {
                "data_template": {
                    "action": "turn_on",
                    "caller": "{{ this.entity_id }}",
                },
                "service": "test.automation",
            },
            "unique_id": "foo-bar-light",
            new_attr: Template(template, hass),
        }
    ] == altered_configs
```